webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/ReportingServiceImpl.scala

```diff
@@ -503,13 +547,77 @@ trait DefaultFindRuleNodeStatusReports extends ReportingService {
       // compute the status
       nodeStatusReports   <- buildNodeStatusReports(runInfos, ruleIds, complianceMode.mode, unexpectedMode)
 
-      t2                  =  System.currentTimeMillis
-      _                   =  TimingDebugLogger.debug(s"Compliance: compute compliance reports: ${t2-t1}ms")
+      t3                  =  System.currentTimeMillis
+      _                   =  TimingDebugLogger.debug(s"Compliance: compute compliance reports: ${t3-t2}ms")
     } yield {
       nodeStatusReports
     }
   }
 
+  override def findRuleNodeCompliance(nodeIds: Set[NodeId], filterByRules : Set[RuleId]): IOResult[Map[NodeId, ComplianceLevel]] = {
+    val t0 = System.currentTimeMillis
+    for {
+      complianceMode      <- getGlobalComplianceMode()
+      unexpectedMode      <- getUnexpectedInterpretation()
+      // we want compliance on these nodes
+      runInfos            <- getNodeRunInfos(nodeIds, complianceMode)
+      t1                  =  System.currentTimeMillis
+      _                   =  TimingDebugLogger.trace(s"Compliance: get node run infos: ${t1-t0}ms")
+
+      // that gives us configId for runs, and expected configId (some may be in both set)
+      expectedConfigIds   =  runInfos.collect { case (nodeId, x:ExpectedConfigAvailable) => NodeAndConfigId(nodeId, x.expectedConfig.nodeConfigId) }
+      lastrunConfigId     =  runInfos.collect {
+        case (nodeId, Pending(_, Some(run), _)) => NodeAndConfigId(nodeId, run._2.nodeConfigId)
+        case (nodeId, x:LastRunAvailable) => NodeAndConfigId(nodeId, x.lastRunConfigId)
+      }
+
+      t2                  =  System.currentTimeMillis
+      _                   =  TimingDebugLogger.debug(s"Compliance: get run infos: ${t2-t0}ms")
+
+      // compute the status
+      nodeStatusReports   <- buildNodeStatusReports(runInfos, filterByRules, complianceMode.mode, unexpectedMode)
+      compliance          = nodeStatusReports.map{ case (k,v) => (k, v.compliance)}
+      t3                  =  System.currentTimeMillis
+      _                   =  TimingDebugLogger.debug(s"Compliance: compute compliance reports: ${t3-t2}ms")
+    } yield {
+      compliance
+    }
+  }.toIO
+
+
+  override def findSystemAndUserRuleCompliances(nodeIds: Set[NodeId], filterBySystemRules : Set[RuleId], filterByUserRules : Set[RuleId]): IOResult[(Map[NodeId, ComplianceLevel], Map[NodeId, ComplianceLevel])] = {
+    val t0 = System.currentTimeMillis
```

Comment:
in a IO for comprehension, you need to use the `currentTimeMillis` in a IO, else you don't know when ZIO actually do the effect